### PR TITLE
ARROW-1281: [C++/Python] Add Docker setup for testing HDFS IO in C++ and Python

### PR DESCRIPTION
**Pull request by wesm**
We aren't testing this in Travis CI because spinning up an HDFS cluster is a bit heavy weight, but this will at least enable us to do easier ongoing validation that this functionality is working properly.

### CODE REVIEW

File: python/testing/hdfs/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# TODO Replace this with a complete clean image build
+FROM cpcloud86/impala:metastore
+
+RUN sudo apt-add-repository -y ppa:ubuntu-toolchain-r/test && \
+    sudo apt-get update && \
```
Comment (xhochy) [inline python/testing/hdfs/Dockerfile]: As you normally run as root inside of docker, sudo should not be required here.

File: python/testing/hdfs/restart_docker_container.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# under the License.
+
+export ARROW_TEST_NN_HOST=arrow-hdfs
+export ARROW_TEST_IMPALA_HOST=$ARROW_TEST_NN_HOST
+export ARROW_TEST_IMPALA_PORT=21050
+export ARROW_TEST_WEBHDFS_PORT=50070
+export ARROW_TEST_WEBHDFS_USER=ubuntu
+
+docker stop $ARROW_TEST_NN_HOST
+docker rm $ARROW_TEST_NN_HOST
```
Comment (xhochy) [inline python/testing/hdfs/restart_docker_container.sh]: Adding `--rm` to `docker run` should make this line redundant

**Review by xhochy — COMMENTED** [review]
I made some initial comments but it seems to be still quite rough so I would wait for more comments until it's settled.

File: python/testing/hdfs/Dockerfile (hunk 1 above)
Comment (wesm) [inline python/testing/hdfs/Dockerfile]: Thanks. I had to change the user back to root from the base image

File: python/testing/hdfs/restart_docker_container.sh (hunk 2 above)
Comment (wesm) [inline python/testing/hdfs/restart_docker_container.sh]: fixed

**Review by wesm — COMMENTED** [review]
Thanks. I'm adding the Python HDFS tests and will remove the WIP

**Comment by wesm** [comment]
This is ready to go. This is good for one-off use but we should see about refactoring our CI scripts to be able to share code more easily with this kind of thing. It's a little bit tricky because of the various interdependent environment variables

As part of ARROW-1213 I will add S3 testing to this setup so you can check things out locally against an access/secret key for a cloud bucket

File: python/testing/hdfs/restart_docker_container.sh (hunk 2 above)
Comment (wesm) [inline python/testing/hdfs/restart_docker_container.sh]: I had some issues when I removed this line and then added `--rm` to the docker run command. Perhaps we can tweak in a subsequent PR when adding more features to these ad hoc Docker tests

**Comment by wesm** [comment]
@xhochy let me know any more comments on this. I'm going to look at the Parquet RC in the meantime

**Comment by wesm** [comment]
+1; I'm going to keep adding some more ad hoc tests

File: python/testing/functions.sh [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+use_gcc() {
+  export CC=gcc-4.9
```
Comment (xhochy) [inline python/testing/functions.sh]: It's better to set these variables in The Dockerfile via `ENV CC gcc-4.9`

File: python/testing/setup_toolchain.sh [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set -e
+
+export PATH="$MINICONDA/bin:$PATH"
+conda update -y -q conda
```
Comment (xhochy) [inline python/testing/setup_toolchain.sh]: in my dockerfiles, i normally include the conda installation in the image. Thus I get faster iteration times on repeated test runs.

File: python/testing/test_hdfs.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set -ex
+
+docker build -t arrow-hdfs-test -f hdfs/Dockerfile .
+bash hdfs/restart_docker_container.sh
```
Comment (xhochy) [inline python/testing/test_hdfs.sh]: For multiple docker containers, have a look at docker-compose. This lets you start and plug multiple containers together.

File: python/testing/setup_toolchain.sh (hunk 4 above)
Comment (xhochy) [inline python/testing/setup_toolchain.sh]: It is also helpful to separate the base installation and the project specific dependencies into different layers of the docker image so they are shared between similar images.

**Review by xhochy — COMMENTED** [review]
Added some comments about best practices I gather from developing a lot with Docker comtainers

File: python/testing/setup_toolchain.sh (hunk 4 above)
Comment (wesm) [inline python/testing/setup_toolchain.sh]: Good point, I will do that in the next patch

File: python/testing/functions.sh (hunk 3 above)
Comment (wesm) [inline python/testing/functions.sh]: I wasn't sure about this as the idea was to be able to easily switch between compilers depending on the test script

File: python/testing/test_hdfs.sh (hunk 5 above)
Comment (wesm) [inline python/testing/test_hdfs.sh]: Will do, thank you

**Pull request closed without merging**